### DISCUSSION TITLE
Package a dependency's compiled exe to be called at runtime

### DISCUSSION THREAD

**Discussion by LewisGaul** [pyinstaller/pyinstaller#6675] [Help] [answered]
I have a python package (https://github.com/LewisGaul/minesolver/tree/main/zig) that consists of a binary executable (compiled Zig code) and a python wrapper that calls it using `subprocess.run()`. Currently this is achieved by packaging the executable as 'package data' (see [here](https://github.com/LewisGaul/minesolver/blob/main/zig/setup.py#L69)) and calling it using `subprocess.run([pathlib.Path(__file__).parent / exe_name, ...])` (see [here](https://github.com/LewisGaul/minesolver/blob/main/zig/zig_minesolver/__init__.py#L33)).

The problem is that when I try to include this dependency in pyinstaller (from [minegauler](https://github.com/LewisGaul/minegauler/blob/main/package/minegauler.spec)) the executable is not found at the `pathlib.Path(__file__).parent / exe_name` location. Perhaps I shouldn't be surprised by this as it's probably not best-practice! I'm not sure whether the binary is getting included in the pyinstaller exe alongside the pyc code (I assume not), but even if I include it in the package using `datas` and `collect_data_files("zig_minesolver")` I'm assuming the python wrapper won't be able to find it...

Any suggestions on how to do this properly? Or, failing that, how to get it to work at all?

**Comment by rokm** [comment, marked as answer]
`subprocess.run([pathlib.Path(__file__).parent / exe_name, ...])` should work, once you ensure that the exe is in fact collected (e.g., via `collect_data_files("zig_minesolver")`, like you suggested).

The compiled .pyc files are collected into executable-embedded PYZ archive while "data files" and "binaries" (i.e., external resources) are collected into (sub-directories in) application's directory. However, in the frozen application, the modules' `__file__` attribute also points to the corresponding sub-directory in application's directory, as if the module was located there. This is done specifically so that external resources can be discovered via the `__file__` attribute. See [docs on this topic](https://pyinstaller.readthedocs.io/en/stable/runtime-information.html#using-file) for details.

**Reply by LewisGaul** [reply]
Thanks, not sure how I missed that in the docs.
